File: check/backtracking/40/40.组合总和Ii.go
package main

import "sort"

/*
 * @lc app=leetcode.cn id=40 lang=golang
 *
 * [40] 组合总和 II
 */

// @lc code=start

/*
解法1 backtrack
*/
func combinationSum2(candidates []int, target int) [][]int {
	if len(candidates) < 1 {
		return nil
	}

	//先进行排序,让相同的元素靠在一起
	sort.Ints(candidates)

	ret := [][]int{}
	var dfs func(record []int, start, sum int)
	dfs = func(record []int, start, sum int) {
		if sum == target {
			ret = append(ret, append([]int{}, record...))
			return
		}
		if sum > target {
			return
		}
		for i := start; i < len(candidates); i++ {
			//剪枝逻辑，值相同的同一层的树枝，只遍历第一条
			if i > start && candidates[i] == candidates[i-1] {
				continue
			}
			record = append(record, candidates[i])
			dfs(record, i+1, sum+candidates[i])
			record = record[:len(record)-1]
		}
		return
	}
	dfs(make([]int, 0, len(candidates)), 0, 0)

	return ret
}

// @lc code=end

File: python3/40.组合总和-ii.py
#
# @lc app=leetcode.cn id=40 lang=python3
#
# [40] 组合总和 II
#

# @lc code=start
from typing import List


class Solution:

    def dfs(self, start: int, sum: int, record: List):
        if sum == self.target:
            # print("record={}".format(record))
            # 最终结果要复制
            self.ret.append(record.copy())
            return
        for i in range(start, self.n):
            # 确保在一种结果当中,下标相同的位置,元素不重复,就能保证最终结果没有重复解
            if i > start and self.candidates[i] == self.candidates[i - 1]:
                continue
            # 剪枝已经不能满足条件的情况
            if sum + self.candidates[i] > self.target:
                continue
            record.append(self.candidates[i])
            self.dfs(i + 1, sum + self.candidates[i], record)
            record.pop()

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        self.n = len(candidates)
        self.candidates = candidates
        self.target = target
        self.ret = []  #保存结果
        self.candidates.sort()  #先排序,使相同数字能挨在一起
        self.dfs(0, 0, [])
        return self.ret


# @lc code=end
